### PR TITLE
8202990: javafx webview css filter property with display scaling

### PR DESCRIPTION
**Pull request by bhaweshkc**
ImageJava.cpp ignores CompositeOperator parameter in drawImage function due to which shadow was getting drawn on top of actual image. apply given composite operator to graphics context before drawing image to fix this issue. another issue is into WCGraphicsPrismContext.java. while blending two layers, applying state to the destination layer was missed due to which image was not getting drawn with right scale in hidpi mode. apply state to fix the issue.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8202990](https://bugs.openjdk.java.net/browse/JDK-8202990): javafx webview css filter property with display scaling

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Arun Joseph](https://openjdk.java.net/census#ajoseph) (@arun-Joseph - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/279/head:pull/279`
`$ git checkout pull/279`

### CODE REVIEW

**Comment by arun-Joseph** [comment]
> With your patch installed, I see a race condition that manifests in one of two ways:

I couldn't reproduce the race condition or crash with the fix.

> I suspect this might be related to the gc.setCompositeOperation call.

This may not be the cause as the compositeOperator is pushed to a RenderQueue which is decoded from the java side.

**Review by arun-Joseph — APPROVED** [review]
The fix and test looks good.

**Comment by kevinrushforth** [comment]
I spent some time this afternoon going over the fix in more detail and doing extensive testing on both Windows and Mac.

I believe the fix is good. Both by inspection and by instrumenting the code, there are no race conditions or other problems that I can see.

The problem appears to be in the test, or else somewhere in the test harness or the SW pipeline exposed by the test. If I run the new CSSTest in a loop on either Mac or Windows, it will crash intermittently. I then reverted your fix, running the new test (which will throw an expected assertion error), and it still crashes intermittently.

My recommendation is to use a system test, similar to what [SVGTest.java](https://github.com/openjdk/jfx/blob/master/tests/system/src/test/java/test/javafx/scene/web/SVGTest.java) does, rather than a unit test in the javafx.web module, which uses `WebPage::paint`.

**Comment by kevinrushforth** [comment]
I believe that the problem noted above is with the test harness, specifically the `WebPageShim::paint` method. I filed [JDK-8252596](https://bugs.openjdk.java.net/browse/JDK-8252596) to track fixing the tests.

So the fix for _this_ PR should avoid using it, as suggested above.

File: tests/system/src/test/java/test/javafx/scene/web/CSSFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import javafx.scene.web.WebView;
+import javafx.stage.Stage;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import test.util.Util;
+
+import static javafx.concurrent.Worker.State.SUCCEEDED;
+import static org.junit.Assert.assertEquals;
```
Comment (arun-Joseph) [inline tests/system/src/test/java/test/javafx/scene/web/CSSFilterTest.java:44]: assertEquals is not used in this test and can be removed.

File: tests/system/src/test/java/test/javafx/scene/web/CSSFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    + " but was:" + colorToString(actual));
+        }
+    }
+
+    private void assertColorNotEquals(String msg, Color notExpected, Color actual, double delta) {
+        double deltaRed = Math.abs(notExpected.getRed() - actual.getRed());
+        double deltaGreen = Math.abs(notExpected.getGreen() - actual.getGreen());
+        double deltaBlue = Math.abs(notExpected.getBlue() - actual.getBlue());
+        double deltaOpacity = Math.abs(notExpected.getOpacity() - actual.getOpacity());
+        if (deltaRed < delta && deltaGreen < delta && deltaBlue < delta && deltaOpacity < delta) {
```
Comment (arun-Joseph) [inline tests/system/src/test/java/test/javafx/scene/web/CSSFilterTest.java:93]: testColorEquals can be reused here.
`if (testColorEquals(expected, actual, delta))`

**Review by arun-Joseph — APPROVED** [review]
The fix and test looks good.

**Review by kevinrushforth — APPROVED** [review]

**Pull request closed without merging**